fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/WorldRenderContext.java

```diff
@@ -0,0 +1,137 @@
+ */
+@Environment(EnvType.CLIENT)
+public interface WorldRenderContext {
+	/**
+	 * The world renderer instance doing the rendering and invoking the event.
+	 *
+	 * @return WorldRenderer instance invoking the event
+	 */
+	WorldRenderer worldRenderer();
+
+	MatrixStack matrixStack();
+
+	float tickDelta();
+
+	long limitTime();
+
+	boolean blockOutlines();
+
+	Camera camera();
+
+	GameRenderer gameRenderer();
+
+	LightmapTextureManager lightmapTextureManager();
+
+	Matrix4f projectionMatrix();
+
+	/**
+	 * Convenient access to {WorldRenderer.world}.
+	 *
+	 * @return world renderer's client world instance
+	 */
+	ClientWorld world();
+
+	/**
+	 * Convenient access to game performance profiler.
+	 *
+	 * @return the active profiler
+	 */
+	Profiler profiler();
+
+	/**
+	 * Test to know if "fabulous" graphics mode is enabled.
+	 *
+	 * <p>Use this for renders that need to render on top of all translucency to activate or deactivate different
+	 * event handlers to get optimal depth testing results. When fabulous is off, it may be better to render
+	 * during {@code WorldRenderLastCallback} after clouds and weather are drawn. Conversely, when fabulous mode is on,
+	 * it may be better to draw during {@code WorldRenderPostTranslucentCallback}, before the fabulous mode composite
+	 * shader runs, depending on which translucent buffer is being targeted.
+	 *
+	 * @return {@code true} when "fabulous" graphics mode is enabled.
+	 */
+	boolean advancedTranslucency();
+
+	/**
+	 * The {@code VertexConsumerProvider} instance being used by the world renderer for most non-terrain renders.
+	 * Generally this will be better for most use cases because quads for the same layer can be buffered
+	 * incrementally and then drawn all at once by the world renderer.
+	 *
+	 * <p>IMPORTANT - all vertex coordinates sent to consumers should be relative to the camera to
+	 * be consistent with other quads emitted by the world renderer and other mods.  If this isn't
+	 * possible, caller should use a separate "immediate" instance.
+	 *
+	 * <p>This property is {@code null} before {@link WorldRenderEvents#BEFORE_ENTITIES} and after
+	 * {@link WorldRenderEvents#BEFORE_DEBUG_RENDER} because the consumer buffers are not available before or
+	 * drawn after that in vanilla world rendering.  Renders that cannot draw in one of the supported events
+	 * must be drawn directly to the frame buffer, preferably in {@link WorldRenderEvents#LAST} to avoid being
+	 * overdrawn or cleared.
+	 */
+	@Nullable VertexConsumerProvider consumers();
+
+	/**
+	 * View frustum, after it is initialized. Will be {@code null} during
+	 * {@link WorldRenderEvents#START}.
+	 */
+	@Nullable Frustum frustum();
+
+	/**
+	 * Used in {@code BLOCK_OUTLINE} to convey the parameters normally sent to
+	 * {@code WorldRenderer.drawBlockOutline}.
+	 */
+	public interface BlockOutlineContext {
```

Comment:
I don't think the `@Environment` propagates to nested types, so this needs it as well.